### PR TITLE
refactor: allow model call to take list as prompt

### PR DESCRIPTION
**Pull request by zenodflow**
Changed logic of when to invoke multiple model calls, allow list as input, and return tool outputs.

### CODE REVIEW

File: libem/core/model.py [hunk 1]
```diff
@@ -107,30 +118,32 @@ def openai(prompt: str, tools: list[str] = None,
...
+                    raise libem.ModelTimedoutException(e)
+
+                response_message = response.choices[0].message
+                tool_calls = response_message.tool_calls
+
+                num_model_calls += 1
+                num_input_tokens += response.usage.total_tokens - response.usage.completion_tokens
+                num_output_tokens += response.usage.completion_tokens
+
+            if num_model_calls == max_model_call:
```
Comment (daiwaid) [inline libem/core/model.py:145]: When max_model_call is reached, the while loop does not terminate, so maybe rename the variable to something like model_calls_warn_threshold?

Comment (zenodflow) [reply libem/core/model.py:146]: Right, sorry, I fixed the bug in a latest commit by setting the tool_calls to []

**Comment by zenodflow** [comment]
One issue we have is how to get the tool calls' output for later processing .. working on it but open to suggestion @daiwaid

**Comment by daiwaid** [comment]
What do you mean by later processing? Can't we just use trace to log it?

**Comment by zenodflow** [comment]
What if we want to _use_ the output from tool, e.g., invoking libem.match as a tool and use its output?

**Comment by daiwaid** [comment]
The easiest solution would just have the LLM write it out, but that'll require an extra pass to the LLM. Alternatively we can have a list of tools that will output the response and if those tools are invoked then just return the response from it directly rather than passing it back to the LLM.

**Comment by zenodflow** [comment]
I see. How about use a separate return value? E.g., in the latest commit

**Comment by zenodflow** [comment]
@daiwaid I'll merge it for now (it should be backward compatible with the rest of the codebase) and can improve for future.

**Pull request merged**